### PR TITLE
[WIP] Standardise the behaviour of the const and non const versions of the array operator within and between the Map classes

### PR DESCRIPTION
**Pull request by madmiraal**
Fixes part 7 of #27638, which was not addressed by #30716.

Part 7 of #27638 points out that in the non const version of the array subscript overloading of the VMap<T, V> template class an instance of V is used uninitialised:
55487295-202fd200-562e-11e9-9230-26c7e63537e4

The non const version of an array subscript overload can be used to both assign a value and give access to a writeable version of an underlying variable. The use of an uninitialised version of V is only a problem when the array subscript is used to access a writeable version of the underlying variable when the array index i.e. the T key is not found in the underlying CowData, but a potential problem nonetheless.

The const version of the VMap array subscript overload calls CRASH_COND when the T key is not found in the underlying CowData.
https://github.com/godotengine/godot/blob/119bf237209414a49879fba40459f22315ab1467/core/vmap.h#L186-L193
This patch mirrors the non const behaviour. Any attempt to use VMap's array subscript to access a value V, when the T key is not found will generate a CRASH_COND instead.

Unfortunately, this will also cause a CRASH_COND if the array subscript is used to assign a V value to a new T value. Since the code was simply calling the public insert function, code using VMap can use this function instead:
https://github.com/godotengine/godot/blob/119bf237209414a49879fba40459f22315ab1467/core/vmap.h#L116-L127
This will have the desired result of overwriting the existing V value if the T key value exists and creating a new pair if it doesn't.

VMap is currently used by Object for Signal, Slot pairs, VisualShader for connections and GDScriptTokenizerBuffer for storing bytes (uint32_t). It appears as if only Object is using VMap's non const version of the array subscript, and it only uses it once for assignment and twice for accessing writeable versions of the underlying Slots. Therefore this patch also makes the following changes:

- Object's use of the array subscript assignment is replaced with the insert function.
- The first use for accessing a writeable version of a Slot is also protected by an initial check to ensure the key exists:
https://github.com/godotengine/godot/blob/119bf237209414a49879fba40459f22315ab1467/core/object.cpp#L1459-L1466
Therefore, no change is required.
- The second use for accessing a writeable version also performs a ERR_FAIL_COND_MSG check, which also returns out of the function.
https://github.com/godotengine/godot/blob/119bf237209414a49879fba40459f22315ab1467/core/object.cpp#L1524-L1526
Therefore, no change is required.

### CODE REVIEW

**Review by RandomShaper — COMMENTED** [review]
I think that following the principle of least astonishment it'd be better to keep the current behavior because that's the same of `std::map`.

Initializing `val` should be enough:
- `V val = V()`;
- or C++11's `V val{}`.

**Comment by madmiraal** [comment]
@RandomShaper I suppose it depends on whether the principle of least astonishment applies to new or existing users of VMap.

Note, this patch also ensures consistent behaviour between the const and non const versions of the array subscript overload. If the behaviour is expected to be consistent and the same as std::map, then the const version, which calls CRASH_COND, would need to be removed. However, there are numerous dependencies on a const version; for example:
https://github.com/godotengine/godot/blob/549aea7284ce8166ba8f5f6fd3f22a03650cb33c/scene/resources/visual_shader.cpp#L1032

I would say the current array subscript overload behaviour (with this patch) is closer to the C++11 [std::map::at](http://www.cplusplus.com/reference/map/map/at/) function; except the std::map::at() function only throws an out_of_range exception instead of aborting when the key is not found. To make the behaviour consistent and the interface the same as std::map, an at() function could be created for VMap and used wherever a const version is required. This would significantly change the VMap interface just to make it the same as std::map, which I don't think is necessary.

That all been said, the aim of this patch is to address the use of an uninitialised variable. Assigning a default value only gets rid of the warning, the variable would still be "uninitialised". In fact, when the values are classes, the original code was already simply calling the default constructor. The problem was only detected because GDScriptTokenizerBuffer uses it to store uint32_t values. Personally, I think the warning should be addressed, not hidden, hence this patch.

**Comment by RandomShaper** [comment]
You are right: `std::map::operator[]` is value-present-or-exception and the right comparison would be with `std::map::at()`. I also understand the new behavior of `VMap::operator[]` is the same whether `const` or not.

So in the end we would be losing a shorter syntax for inserting elements into a `VMap` in exchange for
1. `const`/non-`const` consistency;
2. fixing a warning about an uninitialized variable.

Regarding 1, it's actually a difficult discussion whether that consistency is actually an improvement. Personally, I find the current behavior handy.

But, regarding 2, which is the main point here, I still think that `V val = V();` (or `V val{};`) is enough. That keeps calling the default constructor in case `V` is a class, but zero-initializes if `V` is of a primitive type.

**Comment by madmiraal** [comment]
Since @hpvb so kindly reminded us last week, I can't resist: Is this a case of [Spacebar heating](https://xkcd.com/1172/)?

On a more serious note. This patch has one more additional benefit: It stops VMap imposing a unnecessary requirement on the V type to have a default constructor, since this is the only use of the default constructor in the VMap class.

**Comment by RandomShaper** [comment]
My point is that the main issue can be addressed without those additional changes. They may be beneficial or not, and that can be discussed in a separate PR/proposal.

**Comment by madmiraal** [comment]
Assigning a default value does not address the issue of using an uninitialised variable.

The use of the uninitialised variable or assigning a default value arises when the K key does not exist. This is the same as an index out of bounds in a standard array. We would not want the compiler to extend our array and assign a default value; even if it may be handy to some programmers. We should not want VMap to do so either. When this behaviour is required, VMap has its insert() method.

When the array operator is used to retrieve a value for an index or key that does not exist or is out of bounds, it is not unreasonable for a programme to crash. This is the current behaviour when a const version is requested. It should be the same for a non const version. VMap should not assume that a default value is required just because a non const version was requested.

**Comment by RandomShaper** [comment]
I'm not against your rationale. Sounds good and may be the way to go. In fact, `List` already works that way.

I'm only saying that the uninitialized variable and the interface change are separate concerns.

My suggestion is adding the default-initialization for 3.2 and opening a proposal or PR about the interface consistency for 4.0, which may include additional containers in its scope (`Map`).

**Comment by RandomShaper** [comment]
> Assigning a default value does not address the issue of using an uninitialised variable.

Maybe that is the most important thing to discuss, but I don't quite understand what you mean.

**Comment by madmiraal** [comment]
I agree that the 3.2 milestone should be removed. However, I don't have permissions to change it.

Perhaps a simple example may help with understanding what I mean by:
> Assigning a default value does not address the issue of using an uninitialised variable.

Consider the following VMap<char, int> hexmap:
```
'0' -> 0
'1' -> 1
...
'A' -> 10
...
'F' -> 15
```
With the following code:
```
for (char c = '0'; c < 'F'; ++c) {
    int i = hexmap[c];
    // Do something with i.
}
```
With the current implementation, every attempt to access an invalid key between '9' and 'A' would create a new entry in the hexmap and assign it an arbitrary value. Assigning it a value of 0 instead doesn't stop it from creating new unwanted entries. Note, any future attempts to use hexmap.has() would now be inaccurate. Forcing the programme to crash would draw attention to the problem in this for loop; instead of, for example, wondering why hexmap.has() is not working as intended elsewhere.

**Comment by RandomShaper** [comment]
This PR should do the same change for `Map` and be renamed to something like "Make map classes const/non-const subscript operators consistent". The new name would give a chance to discuss this to other people that may have an opinion but haven't "detected" the kind of change because of the prior name.

And, ideally, there would be another PR for 3.2 just addressing the uninitialized value.

**Comment by madmiraal** [comment]
I see where you're coming from. This problem is bigger than just VMap using an uninitialised variable. Map is assigning a default value, but it also has the same side-effect of creating keys (with default values) when the non const version of the array operator is not used for assignment, but used to retrieve a value for a key that doesn't exist:
https://github.com/godotengine/godot/blob/2d6cf9cc5650318be486f6a7b1f8ad5c74122419/core/map.h#L600-L610

Digging deeper I notice that OrderedHashMap is also creating keys with a default value.
https://github.com/godotengine/godot/blob/2d6cf9cc5650318be486f6a7b1f8ad5c74122419/core/ordered_hash_map.h#L249-L256

HashMap is also creating values:
https://github.com/godotengine/godot/blob/2d6cf9cc5650318be486f6a7b1f8ad5c74122419/core/hash_map.h#L443-L460
but the pair.data value is left uninitialised in the underlying Element:
https://github.com/godotengine/godot/blob/2d6cf9cc5650318be486f6a7b1f8ad5c74122419/core/hash_map.h#L203-L218

Finally, for completion, OAHashMap does not have this problem, because it hasn't overloaded the array operators. If we're looking for consistency, should these be created?

Have I missed any?

**Comment by madmiraal** [comment]
PR #33012 created to just address the uninitialized value in VMap.

**Comment by RandomShaper** [comment]
What a thorough analysis! I don't think there are any cases left.

So on one hand we have the non-const `operator[]`s creating elements. I guess those cases should be given the same crashing behavior for 4.0 (in this PR).

And on the other hand we have "regular" methods, like `HashMap::create_element()`, lacking initialization. Let's have them initializing in 3.2 and 4.0 (in both this and the other PR).

It seems we have a third hand: containers with no indexing operator at all. Well, it's up to you. My suggestion would be doing it for 4.0 in a separate PR on top of this one, or after this one gets merged, to keep the scope of the changes not too big and so this can be merged ASAP.

What do you think?

**Comment by madmiraal** [comment]
I will update PR #33012 to include the initialisation of HashMap's Element's Pair's Tdata to the default value when creating a new element, and I'll begin working on updating this PR to include the standardised behaviour of the array operator between the const and non const versions within and between the Map classes.

**Comment by madmiraal** [comment]
@RandomShaper I see now why you like the way it works at the moment. It allows you to create an indexed set of what I would call singletons. For example from your commit 9692549:

https://github.com/godotengine/godot/blob/2869c13ba2f1ce2b17f30b611eb1eb7f0fa4a472/main/input_default.cpp#L326-L337

The `touch_speed_track` is a `Map` that is used to create and store a `SpeedTrack` singleton associated with a `Ref<InputEventScreenTouch>` `index` that enables the code to track the evolution of a multi-stage touch event (a drag). To avoid a memory leak the singleton is erased when it's no longer needed.

Very handy.

However, I must ask if it is necessary in this specific instance. Would you ever need to simultaneously track multiple, multi-stage touch events? If not and only one SpeedTrack is ever needed, then it may be easier to store the `index` of the `Ref<InputEventScreenTouch>` in a single `SpeedTrack` and reset it when it changes. No `Map` would be needed, and there would be no memory leak in the unlikely event that the end of a touch event is never received.

**Comment by RandomShaper** [comment]
I thought that tracking the speed per pointer made sense, the same way you have a different position for them.

If someone needs an average speed, he can easily compute it, but the other way (forcing users to track themselves the speed individually) would be a hassle.

**Review by RandomShaper — APPROVED** [review]
Thanks for your patience.

**Comment by madmiraal** [comment]
@RandomShaper Thanks for getting back to me. I spent some time looking into why there is a need to track multiple touchscreen events, and I agree that it's necessary.

**Comment by aaronfranke** [comment]
Is this still considered WIP? It needs a rebase.

**Comment by akien-mga** [comment]
After all this time and the rewrite of HashMap, I guess this is not directly relevant as is.

Do you want to keep it open as draft as a reminder to look at it again eventually, or should it be closed for now to revisit later?

**Comment by madmiraal** [comment]
It is still relevant. The inconsistency with and between `Map` classes remains. Sometimes the `operator[]` causes a crash sometimes an insertion.

However, the discussion with @RandomShaper showed that it was a much bigger job than originally identified. As processes have changed, it will now probably need a proposal, and, as it affects core design, it will need @reduz' approval.

At heart is the need to agree not to make `Map`'s array index operator silently create entries as done in `std::map` the number 2 root cause of recurring bugs at Facebook [[1]](https://www.youtube.com/watch?v=lkgszkPnV8g)[[2]](https://www.reddit.com/r/rust/comments/cq9rco/cppcon_2017_curiously_recurring_c_bugs_at_facebook/):
> Bug #2: `std::map`'s index operator will create a default element if the key you search for is not in the map. That's really weird. Two major incidents have occurred at Facebook where the map represented some settings, and printing the settings sneakily inserted new settings with value 0.

**Comment by akien-mga** [comment]
> However, the discussion with @RandomShaper showed that it was a much bigger job than originally identified. As processes have changed, it will now probably need a proposal, and, as it affects core design, it will need @reduz' approval.
> 
> At heart is the need to agree not to make `Map`'s array index operator silently create entries as done in `std::map` the number 2 root cause of recurring bugs at Facebook [[1]](https://www.youtube.com/watch?v=lkgszkPnV8g)[[2]](https://www.reddit.com/r/rust/comments/cq9rco/cppcon_2017_curiously_recurring_c_bugs_at_facebook/):

That makes sense. I'd suggest opening a proposal to discuss this further.

From a quick discussion @reduz seems open to the idea to prevent index operators from silently creating entries if that's possible to do - it's not a behavior we particularly want to rely on, and it does create more issues than it solves.

**Pull request closed without merging**
